pkg/kallsyms/kallsyms.go

```diff
@@ -149,20 +154,34 @@ var (
 //
 //	kallsyms.SpecUpdateAddresses(spec, []string{"socket_file_ops"})
 func SpecUpdateAddresses(spec *ebpf.CollectionSpec, symbols []string) error {
-	kAllSymsFactory := NewKAllSyms
-	return specUpdateAddresses(kAllSymsFactory, spec, symbols)
+	if len(symbols) == 0 {
+		// Nothing to do
+		return nil
+	}
+
+	var err error
+	once.Do(func() {
+		kallsyms, err = NewKAllSyms()
```

Comment:
```
$ cat /proc/kallsyms |wc -c
15861623
```

This file is 16 MiB. In the previous implementation, the kallsyms was released from memory once the gadget using it (profile cpu or top block-io) is no longer using it. In this new implementation, it stays in memory permanently, even when the user stops a gadget such as profile-cpu.

Do you think it is the correct optimization between memory consumption and cpu consumption?

Reply:
I believe the caching logic we have in `specUpdateAddresses` means we can avoid having to keep the content of this file in memory. I wondering if it is better to delay reading this file till we try to resolve the symbols like before, something like this:

```
→ git diff
diff --git a/pkg/kallsyms/kallsyms.go b/pkg/kallsyms/kallsyms.go
index 0598f3a2..0381ff45 100644
--- a/pkg/kallsyms/kallsyms.go
+++ b/pkg/kallsyms/kallsyms.go
@@ -160,16 +160,9 @@ func SpecUpdateAddresses(spec *ebpf.CollectionSpec, symbols []string) error {
                return nil
        }
 
-       once.Do(func() {
-               kallsyms, initErr = NewKAllSyms()
-       })
-       if initErr != nil {
-               return fmt.Errorf("initializing kallsyms: %w", initErr)
-       }
-
        return specUpdateAddresses(
                []symbolResolver{
-                       newKAllSymsResolver(kallsyms),
+                       newKAllSymsResolver(),
                        newEbpfResolver(),
                },
                spec,
diff --git a/pkg/kallsyms/symbol-resolver.go b/pkg/kallsyms/symbol-resolver.go
index e2f74872..48c5dea8 100644
--- a/pkg/kallsyms/symbol-resolver.go
+++ b/pkg/kallsyms/symbol-resolver.go
@@ -32,15 +32,19 @@ type kAllSymsResolver struct {
 }
 
 // newKAllSymsResolver returns a new kAllSymsResolver.
-func newKAllSymsResolver(kAllSyms *KAllSyms) kAllSymsResolver {
-       if kAllSyms == nil {
-               panic("kAllSyms is nil")
-       }
-       return kAllSymsResolver{kAllSyms: kAllSyms}
+func newKAllSymsResolver() kAllSymsResolver {
+       return kAllSymsResolver{}
 }
 
 // Resolve resolves a kernel symbol using the KAllSyms.
 func (r kAllSymsResolver) Resolve(symbol string) (uint64, error) {
+       var err error
+       if r.kAllSyms == nil {
+               r.kAllSyms, err = NewKAllSyms()
+       }
+       if err != nil {
+               return 0, fmt.Errorf("failed to read kallsyms: %w", err)
+       }
        addr, ok := r.kAllSyms.symbolsMap[symbol]
        if !ok {
                return 0, fmt.Errorf("symbol %q was not found in kallsyms", symbol)

```

What do you say?

Reply:
The general idea looks good to me.

One small documentation issue:
```
// Resolve resolves a kernel symbol using the KAllSyms.
func (r kAllSymsResolver) Resolve(symbol string) (uint64, error) {
```

With the test `r.kAllSyms == nil`, this can not be called in parallel from multiple go routine. This is fine because the only user of this API is `specUpdateAddresses()` and it does not do parallel calls. But you can document it.

Btw, can the function `Resolve()` be unexported (renamed to `resolve()`)? Or does Golang impose methods from interfaces to be exported?

Reply:
I updated the code to ensure do don't keep the content of `/proc/kallsyms` in memory